tiledb/sm/serialization/group.cc

```diff
@@ -0,0 +1,293 @@
+#include "tiledb/sm/enums/serialization_type.h"
+#include "tiledb/sm/group/group.h"
+#include "tiledb/sm/group/group_member_v1.h"
+#include "tiledb/sm/misc/constants.h"
+#include "tiledb/sm/serialization/group.h"
+
+#include <set>
+
+using namespace tiledb::common;
+
+namespace tiledb {
+namespace sm {
+namespace serialization {
+
+#ifdef TILEDB_SERIALIZATION
+
+Status group_member_to_capnp(
+    const tdb_shared_ptr<GroupMember>& group_member,
+    capnp::GroupMember::Builder* group_member_builder) {
+  if (group_member == nullptr)
+    return LOG_STATUS(Status_SerializationError(
+        "Error serializing group member; group is null."));
+
+  std::string type_str = object_type_str(group_member->type());
+  group_member_builder->setType(type_str);
+
+  group_member_builder->setUri(group_member->uri().to_string());
+
+  return Status::Ok();
+}
+
+std::tuple<Status, std::optional<tdb_shared_ptr<GroupMember>>>
+group_member_from_capnp(capnp::GroupMember::Reader* group_member_reader) {
+  if (!group_member_reader->hasUri()) {
+    return {Status_SerializationError(
+                "Incomplete group member type in serialization, missing uri"),
+            std::nullopt};
+  }
+
+  if (!group_member_reader->hasType()) {
+    return {Status_SerializationError(
+                "Incomplete group member type in serialization, missing type"),
+            std::nullopt};
+  }
+
+  ObjectType type;
+  RETURN_NOT_OK_TUPLE(
+      object_type_enum(group_member_reader->getType().cStr(), &type),
+      std::nullopt);
+  const char* uri = group_member_reader->getUri().cStr();
+  tdb_shared_ptr<GroupMember> group_member =
+      tdb::make_shared<GroupMemberV1>(HERE(), URI(uri), type);
+
+  return {Status::Ok(), group_member};
+}
+
+Status group_to_capnp(
+    const Group* group, capnp::Group::Builder* group_builder) {
+  if (group == nullptr)
+    return LOG_STATUS(
+        Status_SerializationError("Error serializing group; group is null."));
+
+  const auto& group_members = group->members();
+  if (!group_members.empty()) {
+    auto group_members_builder =
+        group_builder->initMembers(group_members.size());
+    decltype(group_members_builder.size()) i{0};
+    //    uint64_t i = 0;
+    for (const auto& it : group_members) {
+      auto group_member_builder = group_members_builder[i];
+      RETURN_NOT_OK(group_member_to_capnp(it.second, &group_member_builder));
+      // Increment index
+      ++i;
+    }
+  }
+
+  const auto& group_members_to_add = group->members_to_add();
+  if (!group_members_to_add.empty()) {
+    auto group_members_to_add_builder =
+        group_builder->initMembersToAdd(group_members_to_add.size());
+    decltype(group_members_to_add_builder.size()) i{0};
+    for (const auto& it : group_members_to_add) {
+      auto group_member_to_add_builder = group_members_to_add_builder[i];
+      RETURN_NOT_OK(
+          group_member_to_capnp(it.second, &group_member_to_add_builder));
+      // Increment index
```

Comment:
I would remove those increment index comments and add a commend to each section of this function to say what it's doing.